File: thresholding/hsv_segmentation_heavy.py
import sys
from pathlib import Path
from typing import Tuple

import cv2
import numpy as np


def process_images_in_folder_heavy(folder_path: str, resized_size: Tuple[int, int] = (320, 180)) -> None:
    folder = Path(folder_path)
    if not folder.exists() or not folder.is_dir():
        print(f"Invalid folder: {folder_path}")
        return

    # Output directory: <folder>/hsv_segmentation_heavy
    out_dir = folder / "hsv_segmentation_heavy"
    out_dir.mkdir(parents=True, exist_ok=True)

    # HSV green bounds (tunable)
    lower_green = (36, 20, 20)
    upper_green = (145, 255, 255)

    image_paths = [
        p for p in folder.iterdir()
        if p.is_file() and p.suffix.lower() in {".jpg", ".jpeg"}
    ]

    # Pre-create structuring elements
    kernel_open = cv2.getStructuringElement(cv2.MORPH_ELLIPSE, (3, 3))
    kernel_close = cv2.getStructuringElement(cv2.MORPH_ELLIPSE, (5, 5))

    for img_path in image_paths:
        bgr = cv2.imread(str(img_path), cv2.IMREAD_COLOR)
        if bgr is None:
            print(f"Failed to read image: {img_path}")
            continue

        # Resize to working resolution
        small_bgr = cv2.resize(bgr, resized_size, interpolation=cv2.INTER_LINEAR)

        # Denoise while preserving edges
        small_bgr = cv2.bilateralFilter(small_bgr, d=7, sigmaColor=75, sigmaSpace=75)

        # Convert to HSV
        hsv = cv2.cvtColor(small_bgr, cv2.COLOR_BGR2HSV)

        # CLAHE on V channel to normalize lighting and enhance contrast
        h, s, v = cv2.split(hsv)
        clahe = cv2.createCLAHE(clipLimit=2.0, tileGridSize=(8, 8))
        v_eq = clahe.apply(v)
        hsv_eq = cv2.merge((h, s, v_eq))

        # Apply green threshold on equalized HSV
        mask = cv2.inRange(hsv_eq, lower_green, upper_green)

        # Suppress low-saturation areas to avoid false positives
        sat_thresh = 25
        sat_mask = cv2.threshold(s, sat_thresh, 255, cv2.THRESH_BINARY)[1]
        mask = cv2.bitwise_and(mask, sat_mask)

        # Morphological cleanup: open (remove noise) then close (fill gaps)
        mask = cv2.morphologyEx(mask, cv2.MORPH_OPEN, kernel_open, iterations=1)
        mask = cv2.morphologyEx(mask, cv2.MORPH_CLOSE, kernel_close, iterations=1)

        # Remove small connected components (area filter)
        num_labels, labels, stats, _ = cv2.connectedComponentsWithStats(mask, connectivity=8)
        if num_labels > 1:
            clean = np.zeros_like(mask)
            min_area = max(20, int(0.001 * (resized_size[0] * resized_size[1])))
            for label in range(1, num_labels):
                area = stats[label, cv2.CC_STAT_AREA]
                if area >= min_area:
                    clean[labels == label] = 255
            mask = clean

        # Percentage calculation at working scale
        green_pixels = int(cv2.countNonZero(mask))
        percentage = (green_pixels / float(resized_size[0] * resized_size[1])) * 100.0
        rounded = int(round(percentage))

        # Upscale mask to original resolution for output
        mask_upscaled = cv2.resize(mask, (bgr.shape[1], bgr.shape[0]), interpolation=cv2.INTER_NEAREST)

        # Save as <stem>_<percentage>.jpeg in heavy output dir
        stem = img_path.stem
        out_path = out_dir / f"{stem}_{rounded}.jpeg"
        ok = cv2.imwrite(str(out_path), mask_upscaled)
        if not ok:
            print(f"Failed to write: {out_path}")


def main() -> int:
    if len(sys.argv) < 2:
        print(f"Usage: {Path(sys.argv[0]).name} <folder>")
        return 1
    process_images_in_folder_heavy(sys.argv[1])
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: thresholding/hsv_segmentation_heavy.cpp
#include <opencv2/opencv.hpp>
#include <filesystem>
#include <string>
#include <vector>
#include <iostream>
#include <cmath>

namespace canopy_cam {

void process_images_in_folder_heavy(const std::string& folder_path) {
    namespace fs = std::filesystem;
    
    if (!fs::exists(folder_path) || !fs::is_directory(folder_path)) {
        std::cerr << "Invalid folder: " << folder_path << std::endl;
        return;
    }
    
    const cv::Size resized_size(320, 180);
    const cv::Scalar lower_green(36, 20, 20);
    const cv::Scalar upper_green(145, 255, 255);
    
    // Pre-create structuring elements
    cv::Mat kernel_open = cv::getStructuringElement(cv::MORPH_ELLIPSE, cv::Size(3, 3));
    cv::Mat kernel_close = cv::getStructuringElement(cv::MORPH_ELLIPSE, cv::Size(5, 5));
    
    for (const fs::directory_entry& entry : fs::directory_iterator(folder_path)) {
        if (!entry.is_regular_file()) continue;
        
        std::string ext = entry.path().extension().string();
        if (ext != ".jpg" && ext != ".jpeg" && ext != ".JPG" && ext != ".JPEG") continue;
        
        std::string input_path = entry.path().string();
        cv::Mat bgr = cv::imread(input_path, cv::IMREAD_COLOR);
        if (bgr.empty()) {
            std::cerr << "Failed to read image: " << input_path << std::endl;
            continue;
        }
        
        // Resize to working resolution
        cv::Mat small_bgr;
        cv::resize(bgr, small_bgr, resized_size, 0, 0, cv::INTER_LINEAR);
        
        // Denoise while preserving edges
        cv::Mat denoised;
        cv::bilateralFilter(small_bgr, denoised, 7, 75, 75);
        
        // Convert to HSV
        cv::Mat hsv;
        cv::cvtColor(denoised, hsv, cv::COLOR_BGR2HSV);
        
        // CLAHE on V channel to normalize lighting and enhance contrast
        std::vector<cv::Mat> hsv_channels;
        cv::split(hsv, hsv_channels);
        
        cv::Ptr<cv::CLAHE> clahe = cv::createCLAHE(2.0, cv::Size(8, 8));
        clahe->apply(hsv_channels[2], hsv_channels[2]); // Apply to V channel
        
        cv::Mat hsv_eq;
        cv::merge(hsv_channels, hsv_eq);
        
        // Apply green threshold on equalized HSV
        cv::Mat mask;
        cv::inRange(hsv_eq, lower_green, upper_green, mask);
        
        // Suppress low-saturation areas to avoid false positives
        cv::Mat sat_thresh;
        cv::threshold(hsv_channels[1], sat_thresh, 25, 255, cv::THRESH_BINARY);
        cv::bitwise_and(mask, sat_thresh, mask);
        
        // Morphological cleanup: open (remove noise) then close (fill gaps)
        cv::Mat mask_clean;
        cv::morphologyEx(mask, mask_clean, cv::MORPH_OPEN, kernel_open, cv::Point(-1, -1), 1);
        cv::morphologyEx(mask_clean, mask, cv::MORPH_CLOSE, kernel_close, cv::Point(-1, -1), 1);
        
        // Remove small connected components (area filter)
        cv::Mat labels, stats, centroids;
        int num_labels = cv::connectedComponentsWithStats(mask, labels, stats, centroids, 8);
        
        if (num_labels > 1) {
            cv::Mat clean = cv::Mat::zeros(mask.size(), mask.type());
            int min_area = std::max(20, static_cast<int>(0.001 * resized_size.area()));
            
            for (int label = 1; label < num_labels; ++label) {
                int area = stats.at<int>(label, cv::CC_STAT_AREA);
                if (area >= min_area) {
                    clean.setTo(255, labels == label);
                }
            }
            mask = clean;
        }
        
        // Percentage calculation at working scale
        int green_pixels = cv::countNonZero(mask);
        double percentage = static_cast<double>(green_pixels) / static_cast<double>(resized_size.area()) * 100.0;
        int rounded = static_cast<int>(std::round(percentage));
        
        // Upscale mask to original resolution for output
        cv::Mat mask_upscaled;
        cv::resize(mask, mask_upscaled, bgr.size(), 0, 0, cv::INTER_NEAREST);
        
        // Save as <stem>_<percentage>.jpeg in heavy output dir
        std::string stem = entry.path().stem().string();
        fs::path out_dir = entry.path().parent_path() / "hsv_segmentation_heavy";
        fs::create_directories(out_dir);
        
        fs::path out_path = out_dir / (stem + "_" + std::to_string(rounded) + ".jpeg");
        
        if (!cv::imwrite(out_path.string(), mask_upscaled)) {
            std::cerr << "Failed to write: " << out_path.string() << std::endl;
        }
    }
}

} // namespace canopy_cam

int main(int argc, char** argv) {
    if (argc < 2) {
        std::cerr << "Usage: " << argv[0] << " <folder>" << std::endl;
        return 1;
    }
    canopy_cam::process_images_in_folder_heavy(argv[1]);
    return 0;
}
